### PR TITLE
Add example of connecting a Realm React app to sync

### PR DESCRIPTION
**Pull request by tomduncalf**
## What, How & Why?
Update the templates to have the ability to setup sync with a given appId.  
Base all the template source code from the example/ folder
Use the latest `@realm/react`

## ☑️ ToDos

* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 🚦 Tests
* [ ] 📱 Check the React Native/other sample apps work if necessary
* [ ] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: templates/react-native-template-realm-js/template/app/AppWrapper.js
Comment (takameyer) [inline templates/react-native-template-realm-js/template/app/AppWrapper.js:146]: I was getting a very cryptic crash `Illegal Realm Path`, because i set the partition key type to `objectId`.  Maybe we need to document this possible issue. (Would be good if we had a backend template that worked with this automatically)

**Comment by tomduncalf** [comment]
@takameyer I updated this to use flexible sync and to match the new readme format if you have time to take a look, would be nice to get it merged

File: templates/react-native-template-realm-js/template/app/App.js
Comment (takameyer) [inline templates/react-native-template-realm-js/template/app/App.js:1]: I think we need to get rid of these copyright headers, otherwise we might cause worries for developers using the template that they may be violating some law in using it.

File: templates/react-native-template-realm-ts/README.md
Comment (takameyer) [inline templates/react-native-template-realm-ts/README.md:41]: Should we even be providing these steps ourselves?  There has got to be documentation on this in our docs pages.  I think we will have a hard time maintaining this if there are any changes to the UI in Atlas.

File: templates/react-native-template-realm-ts/README.md
Comment (takameyer) [inline templates/react-native-template-realm-ts/README.md:82]: I would love to do this with a script, but we can do that later.

File: templates/react-native-template-realm-ts/template/app/models/Task.ts
Comment (takameyer) [inline templates/react-native-template-realm-ts/template/app/models/Task.ts:25]: Optional arguments should come last.  Otherwise when userId is not defined it will need to be called like this:
```
generate(undefined, "task description")
```

**Review by takameyer — COMMENTED** [review]
Just wanted to provide an initial review.  I'm going to run the template now and get a feel for the configuration.  I'll do a second review if I have anything else to add

File: templates/react-native-template-realm-ts/template/app/App.tsx
Comment (takameyer) [inline templates/react-native-template-realm-ts/template/app/App.tsx:172]:
```suggestion
    backgroundColor: colors.purpleDark,
```

File: templates/react-native-template-realm-ts/template/app/App.tsx
Comment (takameyer) [inline templates/react-native-template-realm-ts/template/app/App.tsx:176]:
```suggestion
    ...buttonStyles.text
```

File: templates/react-native-template-realm-ts/template/app/styles/button.ts
Comment (takameyer) [inline templates/react-native-template-realm-ts/template/app/styles/button.ts:27]:
```suggestion
    marginHorizontal: 10,
```

File: templates/react-native-template-realm-ts/template/app/App.tsx
Comment (takameyer) [reply templates/react-native-template-realm-ts/template/app/App.tsx:176]: Same style as the login button text

File: templates/react-native-template-realm-ts/template/app/App.tsx
Comment (takameyer) [reply templates/react-native-template-realm-ts/template/app/App.tsx:172]: Same styles as the login buttons

File: templates/react-native-template-realm-ts/template/app/styles/button.ts
Comment (takameyer) [reply templates/react-native-template-realm-ts/template/app/styles/button.ts:27]: This makes the button look nice by itself, otherwise its pushed off screen to the left

**Review by takameyer — CHANGES_REQUESTED** [review]
Got a few more comments and suggestions.
I think the anonymous flow is broken.  It still did password authentication for me.

File: templates/react-native-template-realm-js/README.md
Comment (tomduncalf) [reply templates/react-native-template-realm-js/README.md:35]: Yeah that makes sense, also will declutter the TS doc

File: templates/react-native-template-realm-js/template/app/App.js
Comment (tomduncalf) [reply templates/react-native-template-realm-js/template/app/App.js:32]: Yeah I can see what you mean... will mean some duplication but I guess that's fine

File: templates/react-native-template-realm-js/template/app/App.js
Comment (tomduncalf) [reply templates/react-native-template-realm-js/template/app/App.js:40]: Doh

File: templates/react-native-template-realm-js/template/app/AppWrapper.js
Comment (tomduncalf) [reply templates/react-native-template-realm-js/template/app/AppWrapper.js:37]: Yeah that seems reasonable

File: templates/react-native-template-realm-js/template/app/config/sync.js
Comment (tomduncalf) [reply templates/react-native-template-realm-js/template/app/config/sync.js:21]: 😬

File: templates/react-native-template-realm-ts/README.md
Comment (tomduncalf) [reply templates/react-native-template-realm-ts/README.md:41]: I feel like this might be OK as an interim solution, and we can speak to the docs team about creating something easier to follow (like illustrated, or a video).... and I guess give feedback that the cloud setup experience could be a lot easier for first time users

File: templates/react-native-template-realm-ts/template/app/App.tsx
Comment (tomduncalf) [reply templates/react-native-template-realm-ts/template/app/App.tsx:62]: Stupid hooks 🤣

File: templates/react-native-template-realm-ts/template/app/AppWrapper.tsx
Comment (tomduncalf) [reply templates/react-native-template-realm-ts/template/app/AppWrapper.tsx:25]: Split out sync enabled and anonymous auth, keep app id in sync config

File: example/android/app/_BUCK
Comment (tomduncalf) [inline example/android/app/_BUCK:38]: Is this correct or would the more generic `realmreactexample` make more sense if we are deriving all templates from this one?

File: example/android/app/src/main/res/values/strings.xml
Comment (tomduncalf) [inline example/android/app/src/main/res/values/strings.xml:2]:
```suggestion
    <string name="app_name">RealmReactExample</string>
```

File: example/app/AppSync.tsx
Comment (tomduncalf) [inline example/app/AppSync.tsx:12]: Looks like this is unused

File: example/app/AppWrapperSync.tsx
Comment (tomduncalf) [inline example/app/AppWrapperSync.tsx:93]:
```suggestion
          <Pressable style={styles.authButton} onPress={handleLogout}>
            <Text style={styles.authButtonText}>
              Logout {app.currentUser.profile.email}
            </Text>
          </Pressable>
```

File: example/ios/Podfile
Comment (tomduncalf) [inline example/ios/Podfile:6]: Same question re: naming here – not a big deal if it's a pain to change!

File: templates/docs/sync-setup.md
Comment (tomduncalf) [inline templates/docs/sync-setup.md:50]: as is this

File: templates/scripts/sync_rn_ts_to_all_templates.sh
Comment (tomduncalf) [inline templates/scripts/sync_rn_ts_to_all_templates.sh:26]: Could we add a step to delete all the target files first? If everything in `app` is always going to be a Typescript file then like we chatted about, we could just `rm -rf app` then `cp -R` the TS `app` directory into place

File: example/app/models/Task.ts
Comment (kraenhansen) [inline example/app/models/Task.ts:35]: I think it's a bit confusing that the `Task.ts` is exporting this. I think we want to show a pattern that can be extended as new models are added. How about moving this to an `index.ts` and changing it to a named export called `TaskRealmContext` or something.

File: example/app/components/AddTaskForm.tsx
Comment (kraenhansen) [inline example/app/components/AddTaskForm.tsx:15]: In general I've seen props being declared using `type` more often than `interface` [these days](https://stackoverflow.com/a/59561874/503899). In any case I think we should keep this consistent across the components.

File: example/app/AppWrapperSync.tsx
Comment (kraenhansen) [inline example/app/AppWrapperSync.tsx:30]: Is it on purpose we're not using `useCallback` to memoize these arrow functions?

**Review by kraenhansen — APPROVED** [review]
I had a skim over this, as it is a rather large set of changes.
It is to some less critical components though (with all due respect), so I don't see any reason holding this back.

I do however have a few comments and suggested changes.

File: example/app/models/Task.ts
Comment (takameyer) [reply example/app/models/Task.ts:35]: oh yeah...that's a good idea.

File: example/app/AppWrapperSync.tsx
Comment (takameyer) [reply example/app/AppWrapperSync.tsx:30]: Oversight.  Going to change that.  Good catch

**Pull request merged**
